Sources/LSPLogging/SetGlobalLogFileHandler.swift

```diff
@@ -0,0 +1,174 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import RegexBuilder
+
+#if !canImport(os) || SOURCEKITLSP_FORCE_NON_DARWIN_LOGGER
+fileprivate struct FailedToCreateFileError: Error, CustomStringConvertible {
+  let logFile: URL
+
+  var description: String {
+    return "Failed to create log file at \(logFile)"
+  }
+}
+/// Creates a file handle that can write to the log file at `logFilePath`.
+///
+/// If creation of the file handle fails (eg. because the file is not writable), return `nil`.
+private func createLogFileHandle(at logFile: URL) throws -> FileHandle {
+  if !FileManager.default.fileExists(atPath: logFile.path) {
+    guard FileManager.default.createFile(atPath: logFile.path, contents: nil) else {
+      throw FailedToCreateFileError(logFile: logFile)
+    }
+  }
+  let logFileHandle = try FileHandle(forWritingTo: logFile)
+  try logFileHandle.seekToEnd()
+  return logFileHandle
+}
+
+/// The number of log calls that have been made to the log file handler.
+///
+/// See comment in `logImpl`.
+@LogHandlerActor
+fileprivate var logCallsCounter = 0
+
+/// Log the given message to the given log file handle.
+///
+/// Occasionally check if the log file exceeds the target log size. If this is the case, truncate the beginning of the
+/// log to reduce the log size.
+@LogHandlerActor
+private func logToFile(message: String, logFileHandle: FileHandle, logFile: URL, targetLogSize: Int) throws {
+  guard let data = message.data(using: .utf8) else {
+    fputs(
+      """
+      Failed to convert log message to UTF-8 data
+      \(message)
+
+      """,
+      stderr
+    )
+    return
+  }
+  try logFileHandle.write(contentsOf: data)
+  logCallsCounter &+= 1
+
+  // Every `targetLogSize / 10,000` calls, check if the log file exceeds the maximum size. If this is the case, reduce
+  // the log file to 80% of the maximum size by trimming the beginning to allow room for some more log messages.
+  // 10,000 is the maximum log message size (capped by `NonDarwinLogger.log`). This means that the log will never exceed
+  // the target size by more than 80%: The log might be 80% of the maximum size to start with and we might add up to
+  // `targetLogSize` bytes to it until the next check.
+  // For a target log size of 5MB, this means that we check the log for its size every 500 log calls.
+  if logCallsCounter.isMultiple(of: targetLogSize / 10_000),
```

Comment:
Do we need to worry about the case where `targetLogSize < 10_000`, where `isMultiple(of: 0)` will always be false?